src/v/cluster/cluster_recovery_manager.cc

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "cluster/cluster_recovery_manager.h"
+
+#include "cloud_storage/cache_service.h"
+#include "cloud_storage/remote.h"
+#include "cloud_storage/remote_file.h"
+#include "cluster/cloud_metadata/cluster_manifest.h"
+#include "cluster/cloud_metadata/manifest_downloads.h"
+#include "cluster/cluster_recovery_table.h"
+#include "cluster/cluster_utils.h"
+#include "cluster/commands.h"
+#include "cluster/config_frontend.h"
+#include "cluster/errc.h"
+#include "cluster/feature_manager.h"
+#include "cluster/logger.h"
+#include "cluster/security_frontend.h"
+#include "config/configuration.h"
+#include "seastarx.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/sharded.hh>
+
+namespace cluster {
+
+cluster_recovery_manager::cluster_recovery_manager(
+  ss::sharded<ss::abort_source>& sharded_as,
+  ss::sharded<controller_stm>& controller_stm,
+  ss::sharded<features::feature_table>& feature_table,
+  ss::sharded<cloud_storage::remote>& remote,
+  ss::sharded<cluster_recovery_table>& recovery_table,
+  ss::sharded<storage::api>& storage,
+  consensus_ptr raft0)
+  : _sharded_as(sharded_as)
+  , _controller_stm(controller_stm)
+  , _feature_table(feature_table)
+  , _remote(remote)
+  , _recovery_table(recovery_table)
+  , _storage(storage)
+  , _raft0(std::move(raft0)) {}
+
+ss::future<std::optional<model::term_id>>
+cluster_recovery_manager::sync_leader(ss::abort_source& as) {
+    if (!_raft0->is_leader()) {
+        co_return std::nullopt;
+    }
+    auto synced_term = _raft0->term();
+    // Make sure we catch up to the committed offset.
+    const auto committed_offset = _raft0->committed_offset();
```

Comment:
committed offset may be advanced as a result of linearizable barrier, it would be more appropriate to snapshot committed offset after we call `linearizable_barrier`